kernel/src/grouping_iterator.rs

```diff
@@ -0,0 +1,402 @@
+use crate::path::ParsedLogPath;
+use std::iter::Peekable;
+use crate::path::AsUrl;
+use crate::FileMeta;
+
+// type to group the delta log files by version
+// example delta log files and their grouping by version:
+// 1. 0000000001.json
+// 2. 0000000001.checkpoint.parquet
+// 3. 0000000002.json
+// 4. 0000000002.checkpoint.0000000001.0000000002.parquet
+// 5. 0000000003.json
+//
+// The version groups are:
+// 1. [1, 2]
+// 2. [3, 4]
+// 3. [5]
+pub struct VersionGroup<L: AsUrl = FileMeta> {
+    version: u64,
+    files: Vec<ParsedLogPath<L>>,
+}
+
+// Files are implementation of an Iterator that yields ParsedLogPath
+// So it can be any type that implements Iterator<Item = ParsedLogPath<L>>
+// Hence their size is not known at compile time, so we use a Box<dyn Iterator<Item = ParsedLogPath<L>>>
+pub struct VersionGroupingIterator<L: AsUrl = FileMeta> {
+    files: Peekable<Box<dyn Iterator<Item = ParsedLogPath<L>>>>,
```

Comment:
Any particular reason to favor `Box<dyn>` over just making it fully templated?
```suggestion
pub struct VersionGroupingIterator<I, L = FileMeta>
where
    L: AsUrl, 
    I: Iterator<Item = ParsedLogPath<L>>,
{
    files: Peekable<I>,
```

Reply:
As mentioned in the comment, here's my understanding, 
 I used `Box<dyn>` deliberately to support arbitrary iterator types that yield `ParsedLogPath`, rather than fixating upon a concrete compile time type to yield the `ParsedLogPath`. .

Reply:
It's normal for rust code to use templates for that -- the user doesn't normally see or care about the template type because it's obtained by calling helper functions; if the _user_ ever decides _they_ need a generic type they can always cast it as a `&dyn Foo` or store it in a `Box<dyn Foo>`.

Reply:
Got it, makes sense.

Reply:
I can already see this leads to much cleaner and simpler code.